lib/ansible/modules/cloud/amazon/s3_bucket_notification.py

```diff
@@ -0,0 +1,266 @@
+        configs = [cfg.raw for cfg in self.full_config() if cfg.name != desired.raw['Id']]
+        self._upload_bucket_config(configs)
+        return configs
+
+    def _upload_bucket_config(self, config):
+        self.client.put_bucket_notification_configuration(
+            Bucket=self.bucket_name,
+            NotificationConfiguration={
+                'LambdaFunctionConfigurations': config
+            })
+
+
+class Config:
+    def __init__(self, content):
+        self._content = content
+        self.name = content['Id']
+
+    @property
+    def raw(self):
+        return self._content
+
+    def __eq__(self, other):
+        if other:
+            return self.raw == other.raw
+        return False
+
+    @classmethod
+    def from_params(cls, **params):
+        function_arn = params['lambda_function_arn']
+
+        qualifier = None
+        if params['version'] > 0:
+            qualifier = str(params['version'])
+        elif params['alias']:
+            qualifier = str(params['alias'])
+        if qualifier:
+            params['lambda_function_arn'] = '{0}:{1}'.format(function_arn, qualifier)
+
+        return cls({
+            'Id': params['event_name'],
+            'LambdaFunctionArn': params['lambda_function_arn'],
+            'Events': sorted(params['events']),
+            'Filter': {
+                'Key': {
+                    'FilterRules': [{
+                        'Name': 'Prefix',
+                        'Value': params['prefix']
+                    }, {
+                        'Name': 'Suffix',
+                        'Value': params['suffix']
+                    }]
+                }
+            }
+        })
+
+    @classmethod
+    def from_api(cls, config):
+        return cls(config)
+
+
+def main():
+    event_types = ['s3:ObjectCreated:*', 's3:ObjectCreated:Put', 's3:ObjectCreated:Post',
+                   's3:ObjectCreated:Copy', 's3:ObjectCreated:CompleteMultipartUpload',
+                   's3:ObjectRemoved:*', 's3:ObjectRemoved:Delete',
+                   's3:ObjectRemoved:DeleteMarkerCreated', 's3:ObjectRestore:Post',
+                   's3:ObjectRestore:Completed', 's3:ReducedRedundancyLostObject']
+    argument_spec = dict(
+        state=dict(required=False, default='present', choices=['present', 'absent']),
+        event_name=dict(required=True),
+        lambda_function_arn=dict(default=None, aliases=['function_arn']),
+        bucket_name=dict(required=True),
+        events=dict(type='list', default=[], choices=event_types),
+        prefix=dict(required=False, default=''),
+        suffix=dict(required=False, default=''),
+        alias=dict(required=False, default=None),
+        version=dict(type='int', required=False, default=0),
+    )
+
+    module = AnsibleAWSModule(
+        argument_spec=argument_spec,
+        supports_check_mode=True,
+        mutually_exclusive=[['alias', 'version']],
+        required_if=[['state', 'present', ['events', 'lambda_function_arn']]]
```

Comment:
I'd like to remove `lambda_function_arn` in case anyone ever wants to implement other notification types